### DISCUSSION TITLE
What are workspaces in emacs-lsp?  What about multi-root?

### DISCUSSION THREAD

**Discussion by aspiers** [emacs-lsp/lsp-mode#3095] [Q&A] [answered]
Hi, and thanks for this awesome project!  I'm a complete emacs-lsp newbie (since about 2 days ago), despite having used emacs since 1994, but am excited to learn more about it!

I have quite a few questions.  I've tried to find answers in the docs, but am struggling so far.  If people can help me understand then maybe I can submit some PRs to try to clarify these topics and make them easier for the next newbies who arrive :-) 

## Workspaces

What are workspaces in emacs-lsp, exactly? I can't find them defined in [the documentation](https://emacs-lsp.github.io/lsp-mode/).

Does "workspace" refer to the same concept in VS Code?  In other words (if I understand correctly), a workspace is a group of "folders", where a folder is a bit like a project in `projectile.el`?

In which case, is there any equivalent to workspaces in emacs?  I'm not aware of any way to group projectile projects together, or do anything similar.  This leads onto other related questions:

- Does emacs-lsp consider the running emacs process as having a single workspace, or multiple workspaces?  I see `M-x lsp-workspace-folders-{add,remove}` but that doesn't seem to offer a choice of multiple workspaces.
- If multiple, how do I see the workspaces?
- How many language servers can be connected to each (or the) workspace? 

I do appreciate that [LSP was originally designed for VS Code](https://en.wikipedia.org/wiki/Language_Server_Protocol#History), which would result in VS Code terminology like "folders" and "workspaces" making its way into LSP design, and from there into emacs-lsp.  But IMHO it would be really helpful to have a clear explanation of how this terminology maps to emacs.

## Multi-root language servers

I'm also trying to understand the multi-root concept.  Presumably "multi-root language servers" refers to servers which somehow support ["multi-root workspaces"](https://code.visualstudio.com/docs/editor/multi-root-workspaces)?  But as per 2. above, it's not clear what multi-root workspaces would map to in emacs.  Does it mean that each workspace (or the only workspace?) can have one language server (or maybe more?), but each language server can keep track of multiple projects independently, each with its own root directory and own compilation environments, e.g. so that a Javascript module available in one of the roots would not automatically be available in another?

I saw the question in the FAQ "[How do I force lsp-mode to forget the workspace folders for multi root servers so the workspace folders are added on demand?](https://emacs-lsp.github.io/lsp-mode/page/faq/#how-do-i-force-lsp-mode-to-forget-the-workspace-folders-for-multi-root)" but I didn't really understand the question.  In what scenario would you want to forget workspace folders, and what does "added on demand" mean?

Thanks a lot for any light you can shed on any of these questions!

**Comment by yyoncho** [comment, marked as answer]
> Hi, and thanks for this awesome project!  I'm a complete emacs-lsp newbie (since about 2 days ago), despite having used emacs since 1994, but am excited to learn more about it!
>
> I have quite a few questions.  I've tried to find answers in the docs, but am struggling so far.  If people can help me understand then maybe I can submit some PRs to try to clarify these topics and make them easier for the next newbies who arrive :-)
>
> ## Workspaces
>
> What are workspaces in emacs-lsp, exactly? I can't find them defined in [the documentation](https://emacs-lsp.github.io/lsp-mode/).

When I started working on `lsp-mode` it didn't have the notion of workspace folders and it didn't have multi-root support. Unfortunately, the project was using the term `workspace` single connection to the server, so for the `VSCode` term `workspace` we had to use something else so I "invented" the term `session`. So `workspace` in Emacs means `server connection` in VScode.


>
> Does "workspace" refer to the same concept in VS Code?  In other words (if I understand correctly), a workspace is a group of "folders", where a folder is a bit like a project in `projectile.el`?

It is the same in emacs - the session contains group of folders. Note that there is `lsp-load-vscode-workspace` and `lsp-save-vscode-workspace`.

>
> In which case, is there any equivalent to workspaces in emacs?  I'm not aware of any way to group projectile projects together, or do anything similar.  This leads onto other related questions:

It is `session`. Check `M-x lsp-describe-session`. As a side note, we have treemacs integration which will provide similar to VScode experience regarding the sidebar.

>
> - Does emacs-lsp consider the running emacs process as having a single workspace, or multiple workspaces?  I see `M-x lsp-workspace-folders-{add,remove}` but that doesn't seem to offer a choice of multiple workspaces.

It has single `session` with multiple workspaces (server connections).

> - If multiple, how do I see the workspaces?

`M-x lsp-describe-session`

> - How many language servers can be connected to each (or the) workspace?

You can have multiple server connections for one session. You can have multiple server connections for single workspace folder. You can have multiple server connections for single file/buffer. You can have one server managing multiple workspace folders(multi-root),

>
> I do appreciate that [LSP was originally designed for VS Code](https://en.wikipedia.org/wiki/Language_Server_Protocol#History), which would result in VS Code terminology like "folders" and "workspaces" making its way into LSP design, and from there into emacs-lsp.  But IMHO it would be really helpful to have a clear explanation of how this terminology maps to emacs.
>
> ## Multi-root language servers
>
> I'm also trying to understand the multi-root concept.  Presumably "multi-root language servers" refers to servers which somehow support ["multi-root workspaces"](https://code.visualstudio.com/docs/editor/multi-root-workspaces)?  But as per 2. above, it's not clear what multi-root workspaces would map to in emacs.

They map to one server connection managing multiple folders.

> Does it mean that each workspace (or the only workspace?) can have one language server (or maybe more?), but each language server can keep track of multiple projects independently, each with its own root directory and own compilation environments, e.g. so that a Javascript module available in one of the roots would not automatically be available in another?

Usually you have one multi-root connection managing all project of one type(e. g. JDTLS will manage all java projects). Server that does not support multi-root will run one per project(workspace folder).

>
> I saw the question in the FAQ "[How do I force lsp-mode to forget the workspace folders for multi root servers so the workspace folders are added on demand?](https://emacs-lsp.github.io/lsp-mode/page/faq/#how-do-i-force-lsp-mode-to-forget-the-workspace-folders-for-multi-root)" but I didn't really understand the question.  In what scenario would you want to forget workspace folders, and what does "added on demand" mean?

Let say you have java projects A/B. Typically, when you open a file from one of them, both of them will be sent to the server(this is how multi-root works). If you use the snippet and you open a file from A then the server will be started in project A. When you open file from B we won't start new server but we will send `workspaceFolders/didChange` notification and we will manage files from A and B using the same server connection. Note that at any point you can use `M-x lsp-describe-session` and check what is the curret topology.



---

**Reply by aspiers** [reply]
Thanks so much for this very helpful answer!  It certainly clears up a lot of confusion, especially from the incorrect assumption that emacs-lsp workspaces are the same concept as VS Code workspaces.

Can emacs-lsp support multiple sessions, either simultaneously, or maybe just by switching between them?

**Reply by yyoncho** [reply]
> Can emacs-lsp support multiple sessions, either simultaneously, or maybe just by switching between them?

We support switching workspaces. 

>>  It is the same in emacs - the session contains group of folders. Note that there is lsp-load-vscode-workspace and lsp-save-vscode-workspace.

**Reply by aspiers** [reply]
Got it, thanks!

**Reply by caseneuve** [reply]
> Note that there is lsp-load-vscode-workspace and lsp-save-vscode-workspace.

When I use `lsp-load-vscode-workspace` it shuts down the server, but persists the session in a way that the current buffer has no workspace attached to it.  Is that by design?  I would expect this to remove all folders that are not in the saved workspace, but keep/enable workspace for the buffers that belong to the workspace.
